### PR TITLE
refactor datastore

### PR DESCRIPTION
**Pull request by akash1810**
builds on https://github.com/guardian/content_delivery_system/pull/51

refactors Datastore.js to es6 classes for modularity. Uses lodash.template lib to replace the [substituteString function](https://github.com/guardian/content_delivery_system/blob/master/CDS/scripts/js_utils/Datastore.js#L140-L213).

### CODE REVIEW

File: CDS/scripts/js_utils/datastore/config.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+const path = require('path');
+const fs = require('fs');
+const PropertiesReader = require('properties-reader');
+
+class Config {
+    constructor (configDirectory = '/etc/cds_backend/conf.d', namespaceChar = '_') {
```
Comment (fredex42) [inline CDS/scripts/js_utils/datastore/config.js]: a bit concerned that _ is fairly likely to be in a valid key name, so might this break/fail if there are too many _ in the string? e.g., meta_gnm_asset_category for example.

**Comment by fredex42** [comment]
like the use of proper templating but a bit concerned about the _ character, as above

File: CDS/scripts/js_utils/datastore/config.js (hunk 1 above)
Comment (akash1810) [inline CDS/scripts/js_utils/datastore/config.js]: @fredex42 It won't break. The `namespaceChar` is used to key as a key in the `config` property of this class. There isn't a limit on the number of `_`.

**Comment by akash1810** [comment]
I'm going to refactor this even further as I don't think it is necessary to have the substitute string logic as we we're not using it.

**Comment by akash1810** [comment]
Refactored to remove string substitution as it looks to be a complicated abstraction, and we can read and validate the config we need in a simpler way by reading from the db directly each time - this is all substitute string is doing as far as I can tell.

**Comment by akash1810** [comment]
@fredex42 mind reviewing again?

**Comment by fredex42** [comment]
I'm not sure you're right about string substitution - it's used for {meta:*} substitutions as well in order to pick up information passed through from the datastore in a visible and traceable manner

**Comment by akash1810** [comment]
Just re-read my comment and got really confused by what I said 😉 

As far as I can tell, the route file can have a string that looks like `{config:foo}` or `{meta:bar}`. Lets call `config` or `meta` the `prefix`.

If `prefix === 'config'` we attempt to find the key (`foo`) from the object returned from [loadDefs](https://github.com/guardian/content_delivery_system/blob/master/CDS/scripts/js_utils/Datastore.js#L20-L54).

If `prefix !== 'config'` we perform a [database query](https://github.com/guardian/content_delivery_system/blob/master/CDS/scripts/js_utils/Datastore.js#L193-L195).

We don't always use the strings from the route file ([e.g here](https://github.com/guardian/content_delivery_system/blob/master/CDS/scripts/js_utils/youtube/youtube-upload-lib.js#L21-L26)), instead we get values directly from the database.

It would be simpler to always read from the database rather than read from route file, parse the route string then read from the database.

**Comment by akash1810** [comment]
> It would be simpler to always read from the database rather than read from route file, parse/substituteString the route string then read from the database.

Like this - https://github.com/guardian/content_delivery_system/pull/55/files#diff-a53f74ea3d1cb75b89df390b0e0444c3

**Comment by akash1810** [comment]
All PRs are now raised to move media atom cds code to ES6 - https://github.com/guardian/content_delivery_system/pulls?q=is%3Aopen+is%3Apr+label%3AES6

@fredex42 could you take another 👀 please?

File: CDS/scripts/js_utils/datastore/db.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            });
+        });
+    }
+
+    _getOrCreateSource (type) {
+        return new Promise((resolve, reject) => {
+            this._getSource(type)
+                .then(id => resolve(id))
+                .catch(() => {
+                    this._createSource(type)
```
Comment (clloyd) [inline CDS/scripts/js_utils/datastore/db.js]: Does this failing always mean that we should create the source?

Comment (akash1810) [inline CDS/scripts/js_utils/datastore/db.js]: Yeah?? (I'm not sure) Its the [current behaviour](https://github.com/guardian/content_delivery_system/blob/master/CDS/scripts/js_utils/Datastore.js#L60-L71) at least...

**Pull request merged**